### ISSUE TITLE
High-level Event Tracing

### ISSUE THREAD

**Issue by ilia-cher** [pytorch/pytorch#45629]
## 🚀 High-level Event Tracing

Over time we've received requests from community to provide additional information in profiler traces, including:
- High-level events like 'start of a new epoch' or 'another mini-batch iteration'
- Data loader activity, e.g. denoting data loader related time ranges (e.g. reading/processing new minibatch)
- Tracing activity of optimizers

This feature proposal is primarily for the training, driven by the python program.
(We should keep in mind though that some or all of the model can be run by TorchScript runtime)

## Motivation

The motivation is to provide more information in the profiler trace. We currently show three types of ranges in the trace:
- c10/aten ops
- torchscript model methods/functions
- user defined (`record_function` context manager in python/TS) ranges

For all ranges in the trace we save start/end time, (optionally) input arguments shape, and some extra info (e.g. device).

## Pitch

The idea is to add more places inside PyTorch that automatically report themselves to the profiler's trace/stats.

Note: since this is supposed to be used by external tools, the changes proposed here are bound by the backward compatibility requirements.

How this could be implemented:
1. For high-level training loops events: extend `autograd.profiler` context manager by adding a predicate parameter, e.g.
```
   with profiler.profile(
      activities,
      # more args,
      enable_pred =
         enable_when(
            skip_first=20,
            count=5,
            repeat_every=1000,
            warmup=3) as prof:

      for iter in range(N):
         with record_function("iteration " + str(iter)):
            run(minibatch[iter])
         prof.next_minibatch()
```
The exact API may change, but the main idea is to allow users to notify profiler about high-level semantic events, s.t. next minibatch or next epoch using existing `profiler` context manager.
Note on the predicate semantics: predicate param is not strictly required by this proposal, but they work together: the purpose of predicate function is to determine on what iterations to run profiler and on what not to run it (we want to enable profiler only on small % of iterations, but we want a uniform coverage and statistics of training process).
If we add API like `prof.next_minibatch()` this will also allow us to add 'next minibatch' (and also potentially other events like 'next epoch') into the trace.
Note: concepts of 'next minibatch' or 'next epoch' are high-level and ultimately determined by the user, so we allow user to explicitly mark them by e.g. `prof.next_minibatch()`.

2. For dataloaders:

```
for torch.utils.data.IterableDataset.__iter__(): wrap internals into `with record_function(LABEL)`

for torch.utils.data.Dataset. __getitem__(): same
```

Implementation options:
- using predetermined string names ("DataLoader__MyDataLoader_getitem") in `record_function` context manager
- allow `record_function` to take a second param, a user-defined constant (e.g. DATALOADER) in addition to the name of the range (e.g. name of the dataloader class)
- hardcode certain types of user event and subclass `record_function` class (`dataloader_record_function` with DATALOADER hardcoded into)

## Alternatives

There's always an alternative in the form of manual usage of Python's `record_function` context manager: though at the same time, it has to be used explicitly by the user (in the user's model code) to manually label any piece of code by wrapping it in `record_function` context manager.

At the same time, carefully exposing _some_ of the internal events, by default in profiler trace, would make it easier for users, and also external tools, to get a fuller picture of the training process.

## Notes
Some extra functionality not discussed in this doc:
- Automatic registering of eager/script module's 'forward' calls, other eager function calls; tensor-level events.

cc. @gchanan @VitalyFedyunin @ngimel @dzhulgakov @robieta @gdankel

**Comment by dzhulgakov** [comment]
I feel the biggest value for profiling tools would come from semi standardized labels that can be semantically parsed (e.g. how data loading, trainer iteration, eval, etc are called). We can just call them out in the docs, or add some enum in PyTorch. Then all training loop implementations (like Lightning, Ignite or specific model implementation) can opt in to use those labels and work with profiling tools.

Note that you don't really need expressions like `record_function("iteration " + str(iter)):` because one could just refer to the sequential position of the "iteration" label within the parent context.

As for `profiler.profile` it might be better to make the profiler object to be the "condition tracker" and just use the original `with` statement to indicate the iteration. Then you don't need `next_minibatch` and also you won't capture unrelated initialization code. Luckily Python seems to support multi-entrance objects just fine:

```
# once at the beginning
prof = profiler.profile(
     activities,
     enable_pred = enable_when(
            skip_first=20,
            count=5,
            repeat_every=1000,
            warmup=3))
...
      for iter in range(N):
         with prof: # this might log some standard "iteration" mark
            run(minibatch[iter])
```

**Comment by ilia-cher** [comment]
cc. @gchanan  on general frontend API,
@VitalyFedyunin on data loaders

**Comment by gdankel** [comment]
@dzhulgakov I agree that a list of standard labels seems like the right way forward. And downstream tools may provide more advanced capabilities if certain labels and event categories are present, but they should clearly not be required (it would be useful if tools can clearly indicate what the user needs to do in order to enable the extra features though).

I don't have a strong preference regarding the two profiler API variants - the "implicit advance" variant seems pretty elegant to me.

@ilia-cher I'm not sure about adding APIs such as prof.next_minibatch(). Clearly if the original suggestion is used we need a way to "advance" the profiler to allow enable_pred to work. But just having a generic prof.advance() would be preferable I think, and rely on record_function standard labels for marking mini-batches etc.

**Comment by chauhang** [comment]
@ilia-cher @gdankel For training loops like Lightening and other libraries/frameworks building on top of core PyTorch having a semi structured approach with some predefined defaults will be useful. As an example, please see the [Profiler in Lightening](https://pytorch-lightning.readthedocs.io/en/latest/profiler.html)

**Comment by smartcat2010** [comment]
@ilia-cher Besides the above features, it's maybe helpful to consider the following ones(or start another issue?):

1. It's helpful to provide the events on both CPU side and GPU side. For example, if we get a mini-batch iteration's time interval on GPU side, we can analyze the GPU idle time by excluding the GPU kernel busy time intervals.
The "record_function" uses "cudaEventRecord" to make event and "cudaEventElapsedTime" to get the event timestamp. Hopefully it is still kept in this issue's features.

2. Besides c10/aten ops which are in underlying C++ code, many users may care more about the module or function directly called by their Python code. If the Python side code time interval is recorded, it may be more helpful for user to check bottleneck.
For example, nn.Module.forward, torch.Tensor.backward, torch.nn.functional.sigmoid, torch.optim.Optimizer.zero_grad.
Of course, recording events on such fine-grained operations may lead to performance overhead, which also exist on C++ c10/aten ops side.

**Comment by ngimel** [comment]
I agree with @dzhulgakov and @gdankel that we should provide a list of standardized labels. Pytorch does not control training loop abstraction, so we ourselves cannot do much with those labels, but it would provide a way for add-on packages to have an easy semi-standardized way of putting their own labels. 
I'm less certain about implicit advance, for one thing, because it's too magical, for another, because excluding "initialization" is not always desirable or beneficial. Typically the loop looks like
```
for batch in loader.iter():
    model(batch)
```
If `with prof` is below data loop, then profiler will always ignore dataloading which may or may not be desired. 
As for module-level profiling, I remain unconvinced. The actionable bottleneck is always at the function level, pytorch profiler is set up to do profiling at the function level, not module level. With the source attribution that we have now, understanding where the call is coming from is possible, and will bring you to model source. I'm not sure additional infrastructure for module-level profiling actually adds value. If it is indeed required, users can manually instrument their models to add events.

**Comment by ilia-cher** [comment]
> @ilia-cher Besides the above features, it's maybe helpful to consider the following ones(or start another issue?):
> 
> 1. It's helpful to provide the events on both CPU side and GPU side. For example, if we get a mini-batch iteration's time interval on GPU side, we can analyze the GPU idle time by excluding the GPU kernel busy time intervals.
>    The "record_function" uses "cudaEventRecord" to make event and "cudaEventElapsedTime" to get the event timestamp. Hopefully it is still kept in this issue's features.
> 2. Besides c10/aten ops which are in underlying C++ code, many users may care more about the module or function directly called by their Python code. If the Python side code time interval is recorded, it may be more helpful for user to check bottleneck.
>    For example, nn.Module.forward, torch.Tensor.backward, torch.nn.functional.sigmoid, torch.optim.Optimizer.zero_grad.
>    Of course, recording events on such fine-grained operations may lead to performance overhead, which also exist on C++ c10/aten ops side.

thank you for the suggestions, these features are not covered by this issue; also please clarify 1. - I think we already support this - we do have both cpu ranges and (when you set use_cuda=True) cuda kernels ranges - we don't output it in a trace yet but FunctionEvent should have information about cuda kernels within it

**Comment by ilia-cher** [comment]
https://github.com/pytorch/pytorch/pull/47655
